keps/168-pending-workloads-visibility/README.md

```diff
@@ -0,0 +1,655 @@
+	// +optional
+	PendingWorkloadsStatus *LocalQueuePendingWorkloadsStatus
+...
+}
+```
+
+### Cluster Queue API
+
+```golang
+// ClusterQueuePendingWorkload contains the information identifying a pending workload
+// in the cluster queue.
+type ClusterQueuePendingWorkload struct {
+	// Name indicates the name of the pending workload.
+	Name string
+
+	// Namespace indicates the name of the pending workload.
+	Namespace string
+}
+
+type ClusterQueuePendingWorkloadsStatus struct {
+	// TopList contains the list of top pending workloads.
+	// +listType=map
+	// +listMapKey=name
+	// +listMapKey=namespace
+	// +optional
+	TopList []ClusterQueuePendingWorkload
+
+	// LastChangeTime indicates the time of the last change of the structure.
+	LastChangeTime metav1.Time
+}
+
+// ClusterQueueStatus defines the observed state of ClusterQueueStatus
+type ClusterQueueStatus struct {
+...
+	// PendingWorkloadsStatus contains the information exposed about the current
+	// status of the pending workloads in the cluster queue.
+	// +optional
+	PendingWorkloadsStatus *ClusterQueuePendingWorkloadsStatus
+...
+}
+```
+
+### Configuration
+
+```golang
+// Configuration is the Schema for the kueueconfigurations API
+type Configuration struct {
+...
+	// QueueVisibility is configuration to expose the information about the top
+	// pending workloads.
+	QueueVisibility *QueueVisibility
+}
+
+type QueueVisibility struct {
+	// LocalQueues is configuration to expose the information
+	// about the top pending workloads in the local queue.
+	LocalQueues *LocalQueueVisibility
+
+	// ClusterQueues is configuration to expose the information
+	// about the top pending workloads in the cluster queue.
+	ClusterQueues *ClusterQueueVisibility
+
+	// UpdateInterval specifies the time interval for updates to the structure
+	// of the top pending workloads in the queues.
+	// Defaults to 5s.
+	UpdateInterval time.Duration
+}
+
+type LocalQueueVisibility struct {
+	// MaxCount indicates the maximal number of pending workloads exposed in the
+	// local queue status. The maximal value is 1000.
+	// Defaults to 100.
+	MaxCount int32
+}
+
+type ClusterQueueVisibility struct {
+	// MaxCount indicates the maximal number of pending workloads exposed in the
+	// cluster queue status. The maximal value is 1000.
+	// Defaults to 100.
+	MaxCount int32
+}
+```
+
+### In-memory snapshot of the ClusterQueue
+
+In order to be able to quickly compute the top pending workloads per LocalQueue,
+without a need for a prolonged read lock on the ClusterQueue, we create
+periodically in-memory snapshot of the ClusterQueue, organized as a map
+from the LocalQueue to the list of workloads belonging to the ClusterQueue,
+along with their positions. Then, the LocalQueue and ClusterQueue controllers
+do lookup into the cached structure.
+
+The snapshot is taken periodically in intervals configured in
+`QueueVisibility.UpdateInterval`. The snapshot also contains an in-memory
```

Comment:
Since these snapshots are per-cluster queue, we could have a different timer for each ClusterQueue, instead of a single one for all ClusterQueues. That way, we are not holding the lock for too long. And also, we are not sending status updates for all ClusterQueues or LocalQueues at the same time.

Reply:
Good point, and now I'm wondering, we should probably start the periodic snapshot when the cluster queue is created. We also need to delete the snapshot and stop the goroutine when CQ is removed. It makes me think the goroutine should be created and deleted by cluster queue controller.  

Different goroutines will create different Tickers at different moments in time, which should probably be enough.